### Saheed Adisa Ganiyu
### Numerical Method

In [114]:
#importing the necessary modules
import numpy as np
from matplotlib import pyplot as plt
import math

In [115]:
def vec_Pnorm(x, p=2):
    """
    This function computes the p_norm of a vector: 1<=p<=\infty
    
    INPUT:
        
        - `x` vector
        
        - `p` - the type of norm to be computed.
        If `p` is omited, it will be computed. because, it has a default value as 2
        
    OUTPUT:
            -It returns a scalar value.
    """
    if p == float("inf"):
        return max(x)
    x = np.array(x)
    res = (np.sum(x**p)**(1/p))
    return res

### Problem 1

Problem 1 Write a MATLAB or Python program that performs power terations, given a matrix $A$, initial guess vector $x$, tolerance Tol, and maximum number of iterations $N_{\max }$.\
As the stopping criterium let us use the discrepancy between $A x$ and $\lambda x$, i.e. for $x$ normalized so that $\|x\|_2=1$\
(1) compute $\operatorname{Err}=A x-\lambda x$\
(2) if $\|E r r\|_2 \leq$ Tol then exit the iterations.\
Please use the routine you wrote to find the largest eigenvalue and the corresponding eigenvector of the following matrix:
$$
A=\left(\begin{array}{rrrrrr}
-3 & -3 & -2 & -1 & 0 & 1 \\
-3 & 0 & -1 & 0 & 1 & 2 \\
-2 & -1 & 3 & 1 & 2 & 3 \\
-1 & 0 & 1 & 6 & 3 & 4 \\
0 & 1 & 2 & 3 & 9 & 5 \\
1 & 2 & 3 & 4 & 5 & 12
\end{array}\right)
$$
As the initial guess, use $x=(1,2,3,4,5,6)^T$. (Do not forget to normalize it first).\
Let's use Tol $=10^{-8}, N_{\max }=50$. Print out the eigenvalue and eigenvector you found, and the number of iterations it took to converge.\

Organize your code as a function that runs iterations for an arbitrary matrix and initial guess, and a calling script.

#### Solution 1
We implement this algorithm for power iteration.
$$
\begin{aligned}
\text { Algo}&\text{rithm} 27.1. \text{Power Iteration }\\
v^{(0)}= &\text { some vector with }\left\|v^{(0)}\right\|=1\\
\text { for } k&=1,2, \ldots\\
&w=A v^{(k-1)} \quad \text { apply } A\\
&v^{(k)}=w /\|w\| \quad \text { normalize }\\
&\lambda^{(k)}=\left(v^{(k)}\right)^T A v^{(k)} \quad \text { Rayleigh quotient }
\end{aligned}
$$

In [116]:
def pow_iter(A, x, N=50, Tol= 10**(-8)):
    """
    This function finds the largest eigenvalue and its corresponding eigenvector using Power iteration method.
    
    INPUT:
    
        - `A` a sqaure matrix.
        - `x` the initial guess vector.
        - `N` maximum number of iteration.
        - `Tol` tolerance for convergence.
    OUTPUT:
    
        - `lamda_k` the largest Eigenvalue.
        - `vk` the corresponding Eigenvector.
        - `count` the number of iteration for convergence.
    """
    N_max = N
    x = np.array(x)
    vk = x/vec_Pnorm(x)
    count = 0
    Err = 1000
    Tol = Tol
    lamda_k = 1
    while (Err> Tol and count <= N_max):            # Checking for convergence
        w = np.array(np.matmul(A,vk)).ravel()       # Applying A
        vk = w/vec_Pnorm(w)                         # Normalizing
        lamda_k = np.array(np.matmul(np.matmul(vk.T,A),vk)).ravel()  # Rayleigh quotient
        count = count + 1                           # Num of Iteration counter
        Err = (np.matmul(A,vk)) - (lamda_k*vk)
        Err = vec_Pnorm(Err)                        # Error norm
        #print(Err)
    return vk, lamda_k[0], count

In [118]:
# Applying the given matrix and the initial guess vector
A = np.mat([[-3,-3,-2,-1,0,1],[-3,0,-1,0,1,2],[-2,-1,3,1,2,3],[-1,0,1,6,3,4],[0,1,2,3,9,5],[1,2,3,4,5,12]])
x = [1, 2, 3, 4, 5, 6]
v,l,i=pow_iter(A, x)
print("A=\n",A)
print("\n Initial_Vector= ",x)
print("\n===================Results======================================")
print("\n Eigen_Vector= ",v)
print("\n Eigen_Value= ",l)
print("\n Num_iter= ",i)
print("==================================================================")

A=
 [[-3 -3 -2 -1  0  1]
 [-3  0 -1  0  1  2]
 [-2 -1  3  1  2  3]
 [-1  0  1  6  3  4]
 [ 0  1  2  3  9  5]
 [ 1  2  3  4  5 12]]

 Initial_Vector=  [1, 2, 3, 4, 5, 6]

===================Results======================================

 Eigen_Vector=  [-0.01724002  0.0957202   0.22290849  0.36719369  0.53227151  0.72298552]

 Eigen_Value=  18.878503170086887

 Num_iter=  18


**Result:** from the given matrix $A$ and the initial vector $x$, we our largest eigenvalue $= 18.878503170086887$ and its corresponding eigenvector $= [-0.01724002, \quad  0.0957202, \quad  0.22290849, \quad 0.36719369, \quad 0.53227151, \quad 0.72298552]$ with observed number of iteration $=18$ 

### Problem 2

Write a MATLAB program (a function) that performs inverse iteration with a FIXED shift $\mu$, on a given matrix $A$ and given initial guess vector $x$, tolerance Tol, and a maximum number of iterations $N_{\max }$.

As the stopping criterium let us use the discrepancy between $A x$ and $\lambda x$, i.e. for $x$ normalized so that $\|x\|_2=1$\
(1) compute $E r r=A x-\lambda x$\
(2) if $\|E r r\|_2 \leq$ Tol then exit the iterations.\
Set shift $\mu$ to $3.7$.\
Compute the inverse $B$ of $A-\mu I$. (You are allowed to use MATLAB function "inv()" or "numpy.linalg.inv()" in Python for this).

Run you power method iterations on matrix $B$, and find the eigenvalue of $A$ closest to 3.7, and the corresponding eigenvector.

As the initial guess, use $x=(1,2,3,4,5,6)^T$. (Do not forget to normalize it first). Let's use Tol $=10^{-8}, N_{\max }=50$.

Organize your code as a function that runs iterations for an arbitrary matrix and initial guess, and a script that calls your function. Print out the eigenvalue and eigenvector you found, and the number of iterations it took to converge.

#### Solution 2
Here we implement the following algorithm.
$$
\begin{aligned}
\text { Algo}&\text{rithm 27.2. Inverse Iteration }\\
v^{(0)}=&\text { some vector with }\left\|v^{(0)}\right\|=1\\
\text { for } k=&1,2, \ldots\\
&\text { Solve }(A-\mu I) w=v^{(k-1)} \text { for } w \quad \text { apply }(A-\mu I)^{-1}\\
&v^{(k)}=w /\|w\| \quad \text { normalize }\\
&\lambda^{(k)}=\left(v^{(k)}\right)^T A v^{(k)} \quad \text { Rayleigh quotient }
\end{aligned}
$$

In [119]:
def inv_iter(A, x, mu = 3.7, N=100, Tol= 10**(-8)):
    """
    This function finds the eigenvalue of matrix A closest to a given fixed point mu 
    and its corresponding eigenvector using inverse iteration method.
    
    INPUT:
    
        - `A` a sqaure matrix.
        - `x` the initial guess vector.
        - `mu` a fixed shift point.
        - `N` maximum number of iteration.
        - `Tol` tolerance for convergence.
    OUTPUT:
    
        - `lamda_k` the closest eigenvalue to the given fixed shift point mu.
        - `vk` the corresponding Eigenvector.
        - `count` the number of iteration for convergence.
    """
    N_max = N
    x = np.array(x)
    vk = x/vec_Pnorm(x)
    I = np.eye(np.shape(A)[0])
    count = 0
    Err = 1000
    Tol = Tol
    lamda_k = 1
    B = np.linalg.inv(A - (mu*I))
    while (Err> Tol and count <= N_max):                # Checking for convergence
        w = np.array(np.matmul(B,vk)).ravel()           # Applying A
        vk = w/vec_Pnorm(w)                             # Normalizing
        lamda_k = np.array(np.matmul(np.matmul(vk.T,A),vk)).ravel()  # Rayleigh quotient
        count = count + 1                               # Num of Iteration counter
        Err = (np.matmul(A,vk)) - (lamda_k*vk)
        Err = vec_Pnorm(Err)                            # Error norm
    return vk, lamda_k[0], count
    

**2) Part_a:** Here, we apply inverse iteration on matrix $A$

In [121]:
# Applying inverse iteration method on the given matrix A and the initial guess vector x
A = np.mat([[-3,-3,-2,-1,0,1],[-3,0,-1,0,1,2],[-2,-1,3,1,2,3],[-1,0,1,6,3,4],[0,1,2,3,9,5],[1,2,3,4,5,12]])
x = [1, 2, 3, 4, 5, 6]
v,l,i=inv_iter(A, x)
print("A=\n",A)
print("\n Initial_Vector= ",x)
print("\n===================Results===================================")
print("\n Eigen_Vector= ",v)
print("\n Eigen_Value= ",l)
print("\n Num_iter= ",i)
print("==============================================================")

A=
 [[-3 -3 -2 -1  0  1]
 [-3  0 -1  0  1  2]
 [-2 -1  3  1  2  3]
 [-1  0  1  6  3  4]
 [ 0  1  2  3  9  5]
 [ 1  2  3  4  5 12]]

 Initial_Vector=  [1, 2, 3, 4, 5, 6]

===================Results===================================

 Eigen_Vector=  [-0.11100168 -0.09131111 -0.03652608  0.91529329 -0.30140393 -0.22226251]

 Eigen_Value=  4.12214702965213

 Num_iter=  31


**Observation:** from the given matrix $A$ and the initial vector $x$, we have our largest eigenvalue $= 4.12214702965213
$ and its corresponding eigenvector $= [-0.11100168, \quad  -0.09131111, \quad  -0.03652608, \quad 0.91529329, \quad -0.30140393, \quad -0.22226251]$ with observed number of iteration $=31$ 

**2) Part_b:** Here, we apply power iteration on inverse of matrix $B = A - \mu I$

In [122]:
# Applying power iteration on the inverse of matrix 𝐵=𝐴−𝜇𝐼
A = np.mat([[-3,-3,-2,-1,0,1],[-3,0,-1,0,1,2],[-2,-1,3,1,2,3],[-1,0,1,6,3,4],[0,1,2,3,9,5],[1,2,3,4,5,12]])
x = [1, 2, 3, 4, 5, 6]
I = np.eye(np.shape(A)[0])
mu = 3.7
B = np.linalg.inv(A - (mu*I))
v,l,i=pow_iter(B, x)
print("B=\n",B)
print("\n Initial_Vector= ",x)
print("\n===================Results==============================================")
print("\n Eigen_Vector= ",v)
print("\n Eigen_Value= ",l)
print("\n Num_iter= ",i)
print("===========================================================================")

B=
 [[-2.42532937e-01  1.51930637e-01  2.44861565e-01 -2.81747029e-01
   1.81144473e-03  3.87973326e-02]
 [ 1.51930637e-01 -4.14472093e-01  2.57974518e-01 -2.19222949e-01
   3.77295336e-02  7.12450748e-02]
 [ 2.44861565e-01  2.57974518e-01 -1.12001838e+00  2.19413627e-02
   1.76270734e-01  1.96400652e-01]
 [-2.81747029e-01 -2.19222949e-01  2.19413627e-02  1.98867693e+00
  -6.08796067e-01 -5.12814284e-01]
 [ 1.81144473e-03  3.77295336e-02  1.76270734e-01 -6.08796067e-01
   5.83162980e-01 -1.30929319e-01]
 [ 3.87973326e-02  7.12450748e-02  1.96400652e-01 -5.12814284e-01
  -1.30929319e-01  3.53664373e-01]]

 Initial_Vector=  [1, 2, 3, 4, 5, 6]

===================Results==============================================

 Eigen_Vector=  [-0.11100168 -0.09131111 -0.03652607  0.91529329 -0.30140393 -0.22226251]

 Eigen_Value=  2.3688429143373346

 Num_iter=  33


**Observation:** The outcome here aserts the fact that the eigenvectors of $B = (A-\mu I)^{-1}$ are the same as the eigenvectors of $A$, and the corresponding eigenvalues are $\left\{\left(\lambda_j-\mu\right)^{-1}\right\}$, where $\left\{\lambda_j\right\}$ are the eigenvalues of $A$, for any $\mu \in \mathbb{R}$. As we can see the result in **part_a** and  **part_b** of solution (2) where their eigenvector are the same and corresponding eigenvalue is indeed $\left(\lambda_j-\mu\right)^{-1} = 2.3688429143373346$ which shows that $\lambda_j$ and $\mu$ are not closed to each other as the resulted eigenvalue is not large.


### Problem 3.
Use the same algorithm and the same matrix as in Problem 2, but with shift $\mu=$ 4.0. PLEASE COMMENT ON how the change of shift $\mu$ has changed the number of iterations (comparing to problem 2) and why?

**3) Part_a:** Here, we apply inverse iteration on matrix $A$ with $\mu = 4.0$

In [123]:
A = np.mat([[-3,-3,-2,-1,0,1],[-3,0,-1,0,1,2],[-2,-1,3,1,2,3],[-1,0,1,6,3,4],[0,1,2,3,9,5],[1,2,3,4,5,12]])
x = [1, 2, 3, 4, 5, 6]
v,l,i=inv_iter(A, x, mu=4.0 )
print("A=\n",A)
print("\n Initial_Vector= ",x)
print("\n===================Results==============================================")
print("\n Eigen_Vector= ",v)
print("\n Eigen_Value= ",l)
print("\n Num_iter= ",i)
print("==========================================================================")

A=
 [[-3 -3 -2 -1  0  1]
 [-3  0 -1  0  1  2]
 [-2 -1  3  1  2  3]
 [-1  0  1  6  3  4]
 [ 0  1  2  3  9  5]
 [ 1  2  3  4  5 12]]

 Initial_Vector=  [1, 2, 3, 4, 5, 6]

===================Results==============================================

 Eigen_Vector=  [-0.11100168 -0.09131111 -0.03652607  0.91529329 -0.30140393 -0.22226251]

 Eigen_Value=  4.12214702965213

 Num_iter=  9


**Observation:** from the given matrix $A$ and the initial vector $x$, we have our largest eigenvalue $= 4.12214702965213
$ and its corresponding eigenvector $= [-0.11100168, \quad  -0.09131111, \quad  -0.03652607, \quad 0.91529329, \quad -0.30140393, \quad -0.22226251]$ with observed number of iteration $=9$. 

**Remark:** Here, we noticed that, as $\mu = 4.0$ gets closer to corresponding eigenvalue$=4.12214702965213$. So, the rate of convergence becomes faster that is, from $33$ in solution$(2)$ to $9$  here.

**3) Part_b:** Here, we apply power iteration on inverse of matrix $B = A - \mu I$ with $\mu = 4.0$

In [125]:
# Applying power iteration on the inverse of matrix 𝐵=𝐴−𝜇𝐼
A = np.mat([[-3,-3,-2,-1,0,1],[-3,0,-1,0,1,2],[-2,-1,3,1,2,3],[-1,0,1,6,3,4],[0,1,2,3,9,5],[1,2,3,4,5,12]])
x = [1, 2, 3, 4, 5, 6]
I = np.eye(np.shape(A)[0])
mu = 4.0
B = np.linalg.inv(A - (mu*I))
v,l,i=pow_iter(B, x)
print("B=\n",B)
print("\n Initial_Vector= ",x)
print("\n===================Results==============================================")
print("\n Eigen_Vector= ",v)
print("\n Eigen_Value= ",l)
print("\n Num_iter= ",i)
print("==========================================================================")

B=
 [[-0.13333333  0.2         0.2        -0.86666667  0.2         0.2       ]
 [ 0.2        -0.3         0.2        -0.7         0.2         0.2       ]
 [ 0.2         0.2        -0.8        -0.2         0.2         0.2       ]
 [-0.86666667 -0.7        -0.2         6.86666667 -2.2        -1.7       ]
 [ 0.2         0.2         0.2        -2.2         1.2         0.2       ]
 [ 0.2         0.2         0.2        -1.7         0.2         0.7       ]]

 Initial_Vector=  [1, 2, 3, 4, 5, 6]

===================Results==============================================

 Eigen_Vector=  [-0.11100169 -0.09131111 -0.03652607  0.91529329 -0.30140393 -0.22226251]

 Eigen_Value=  8.186854832638703

 Num_iter=  10


**Observation:** The outcome here aserts the fact that the eigenvectors of $B = (A-\mu I)^{-1}$ are the same as the eigenvectors of $A$, and the corresponding eigenvalues are $\left\{\left(\lambda_j-\mu\right)^{-1}\right\}$, where $\left\{\lambda_j\right\}$ are the eigenvalues of $A$, for any $\mu \in \mathbb{R}$. As we can see the result in **part_a** and  **part_b** of solution (3) where their eigenvector are the same and corresponding eigenvalue is indeed $\left(\lambda_j-\mu\right)^{-1} = 8.186854832638703$ which shows that $\lambda_j$ and $\mu$ are not too closed to each other as the resulted eigenvalue is not too large. But it is closer than $\mu = 3.7$ in problem(2) as we can see from the  fast rate of convergence which actually reduced the number of iteration to $10$ from $33$.\
Moreover, when we chose our $\mu = 4.12$ very closed to $\lambda_j$ we have the corresponding eigenvalue $=465.7597527859305$ with number of iteration$=4$ which shows their closeness as it reflects in the rate of convergence and the resulted eigenvalue.


### Problem 4

Given
$$
\left\{\begin{array}{ll}
30 x_1-10+10 \cos \left(0.1 x_2+1\right) & =0 \\
20 x_2+15+5 \sin \left(0.1 x_1+0.1 x_3\right) & =0 \\
70 x_3-30+5 \mathrm{e}^{-x_1^2} & =0
\end{array} .\right.
$$

**(b)** Write a program in Python or Matlab that finds the solution of this system by a simple iteration with $g(x)=x-A f(x)$ where $A$ is either a constant or a constant matrix of your choice, and $f(x)$ is the left hand side of the above system. The initial guess should be $x^{(0)}=(1,1,1)$ and the iterations should be stopped when the condition $\left\|f\left(x^{(\nu)}\right)\right\|_2 \leq 10^{-6}$ is satisfied.

Please, print out the number of iterations required to reach the desired accuracy. Submit the code and the printout.

In [126]:
# Building the given function in question 4(a)
def f(x):
    x = np.array(x)
    f1 = 30*x[0] - 10 + 10*np.cos(0.1*x[1]+1)
    f2 = 20*x[1] + 15 + 5*np.sin(0.1*x[0] + 0.1*x[2])
    f3 = 70*x[2] - 30 + 5*np.exp(-(x[0])**2)
    return np.array([f1,f2,f3])

In [127]:
def simple_iter(x,Nmax=50,Tol=10**(-6)):
    """
    This function find roots of non linear equation using simple iteration method
    
    INPUT:
        - `x` the initial guess vector.
        - `Nmax` maximum number of iteration.
        - `Tol` tolerance for convergence.
    OUTPUT:
        - `xn` the list of solution.
        - `count` the number of iteration for convergence.
    """
    Tol = Tol
    Nmax = Nmax
    x0 = np.array(x)
    Err = 10; count =0
    xn = np.array([1,1,1])
    A = 0.02*np.eye(3)
    while (Err>Tol and count<Nmax):
        xn = x0- np.matmul(A,f(x0))
        x0 =xn
        Err = vec_Pnorm(f(xn))
        count +=1
    return xn, count

In [128]:
# executing the given initial guess
x0 = [1,1,1]
x1,c1=simple_iter(x0, Nmax=100)
print("\n===================Results==========================")
print("Sol= ",x1)
print("\nNum_iter= ",c1)
print("=====================================================")


===================Results==========================
Sol=  [ 0.13239547 -0.76226451  0.35838399]

Num_iter=  34


**Observation:** here we chose our matrix $A$ to be a $3x3$ diagonal matrix with all its entries eqaul $0.02$. Then we have the solution to be $[ 0.13239547\quad -0.76226451\quad  0.35838399]$ with number of iteration equal $34$.

**Problem 5 (a)** Write a program in Matlab that finds the solution of the system below by a simple iteration with $g(x)=x-A f(x)$ where $A$ is either a constant or a constant matrix of your choice, and $f(x)$ is the left hand side of the system:
$$
\left\{\begin{array}{l}
x_1-x_2-10+10 \cos \left(0.1 x_1\right)=0 \\
2 x_1+x_3-20+10 \sin \left(0.1 x_2\right)=0 \\
x_1-0.1 x_2+x_3-30+10 \mathrm{e}^{-x_3^2}=0
\end{array} .\right.
$$
The initial guess should be $x^{(0)}=(1,1,1)$ and the iterations should be stopped when the condition $\left\|f\left(x^{(\nu)}\right)\right\|_2 \leq 10^{-6}$ is satisfied. You do not have to prove anything, just find $A$ that makes the iteration converge. Please, print out the number of iterations required to reach the desired accuracy. Submit the code and the printout.\
**(b)** Numerically solve the system from part (a) using the Newton's method, with the same initial guess and stopping criterion. Please, print out the number of iterations required to reach the desired accuracy. Submit the code and the printout.

**(a)**

In [129]:
# Building the given function in question 5(a)
def f2(x):
    x = np.array(x)
    f1 = x[0] - x[1] - 10 + 10*np.cos(0.1*x[0])
    f2 = 2*x[0] + x[2] - 20 + 10*np.sin(0.1*x[1])
    f3 = x[0] - 0.1*x[1] + x[2] - 30 + 10*np.exp(-(x[2])**2)
    return np.array([f1,f2,f3])

In [130]:
def simple_iter2(x,Nmax=50,Tol=1e-6):
    """
    This function find roots of non linear equation using simple iteration method
    
    INPUT:
        - `x` the initial guess vector.
        - `Nmax` maximum number of iteration.
        - `Tol` tolerance for convergence.
    OUTPUT:
        - `xn` the list of solution.
        - `count` the number of iteration for convergence.
    """
    Tol = Tol
    Nmax = Nmax
    x0 = np.array(x)
    Err = 10; count =0
    xn = np.array([1,1,1])
    A =  0.45*np.eye(3)
    while (Err>Tol and count<Nmax):
        xn = x0- np.matmul(A,f2(x0))
        x0 =xn
        Err = vec_Pnorm(f2(xn))
        count = count + 1
    return xn, count

In [131]:
# executing the given initial guess
x0 = [-4,-5,33]
x2,c2=simple_iter2(x0, Nmax=100)
print("\n===================Results==========================")
print("Sol= ",x2)
print("\nNum_iter= ",c2)
print("=====================================================")


===================Results==========================
Sol=  [-4.38489108 -5.33094986 33.8517956 ]

Num_iter=  44


**Observation:** here we carefully chose our matrix $A$ to be a $3x3$ diagonal matrix with all its entries eqaul $0.45$. Then we have the solution to be $[ -4.38489108\quad -5.33094986\quad  33.8517956]$ with number of iteration equal $44$.

**(b)**

In [132]:
# Building the given function in question 5(a)
def f3(x):
    x = np.array(x)
    f1 = x[0] - x[1] - 10 + 10*np.cos(0.1*x[0])
    f2 = 2*x[0] + x[2] - 20 + 10*np.sin(0.1*x[1])
    f3 = x[0] - 0.1*x[1] + x[2] - 30 + 10*np.exp(-(x[2])**2)
    return np.array([f1,f2,f3])

In [133]:
# Building the corresponding Jacobian matrix
def j(x):
    x = np.array(x)
    jacb = np.array([[1-np.sin(0.1*x[0]), -1, 0],[2, np.cos(0.1*x[1]), 1], [1, -0.1, 1-20*x[2]*np.exp(-(x[2])**2)]])
    return jacb

In [134]:
def newton_methd(x,Nmax=50,Tol=1e-6):
    """
    This function find roots of non linear equation using Newton iteration method
    
    INPUT:
        - `x` the initial guess vector.
        - `Nmax` maximum number of iteration.
        - `Tol` tolerance for convergence.
    OUTPUT:
        - `xn` the list of solution.
        - `count` the number of iteration for convergence.
    """
    Tol = Tol
    Nmax = Nmax
    x0 = np.array(x)
    Err = 10; count =0
    xn = np.array([1,1,1])
    while (Err>Tol and count<Nmax):
        temp = np.linalg.inv(j(x0))
        xn = x0- np.matmul(temp,f2(x0))
        x0 =xn
        Err = vec_Pnorm(f3(xn))
        count +=1
    return xn, count

In [137]:
# executing the given initial guess
x0 = [1,1,1]
x3,c3=newton_methd(x0, Nmax=100)
print("\n===================Results======================")
print("Sol= ",x3)
print("\nNum_iter= ",c3)
print("=================================================")


===================Results======================
Sol=  [-4.38489174 -5.33095023 33.85179672]

Num_iter=  6


**Observation:** with Newton method we can see a pretty improvement on the rate of cinvergence as the number iteration jumped to $6$ instead of $44$ in part_a of simple iteration in problem(5). Therefore, our solution is $[-4.38489174\quad -5.33095023\quad 33.85179672]
$.